packages/flutter/test/widgets/overlay_2_test.dart

```diff
@@ -0,0 +1,1499 @@
+
+  AbstractNode? node = descendant.parent;
+  while (node != null) {
+    if (node is! RenderConstrainedLayoutBuilder<BoxConstraints, RenderBox>) {
+      node = node.parent;
+    } else {
+      node.markNeedsBuild();
+      return;
+    }
+  }
+  assert(false);
+}
+
+void verifyTreeIsClean() {
+   final RenderObject renderObject = RendererBinding.instance.renderView;
+   bool hasDirtyNode = renderObject.debugNeedsLayout;
+
+   void visitor(RenderObject renderObject) {
+     expect(renderObject.debugNeedsLayout, false, reason: '$renderObject is dirty');
+     hasDirtyNode = hasDirtyNode || renderObject.debugNeedsLayout;
+     if (!hasDirtyNode) {
+       renderObject.visitChildren(visitor);
+     }
+   }
+   visitor(renderObject);
+}
+
+void verifyOverlayChildReadyForLayout(GlobalKey overlayWidgetKey) {
+  final RenderObjectElement element = overlayWidgetKey.currentContext! as RenderObjectElement;
+  final RenderBox layoutSurrogate = element.renderObject as RenderBox;
+  assert(element.runtimeType.toString() == '_OverlayPortalElement');
+
+  if (layoutSurrogate.debugNeedsLayout) {
+    assert(layoutSurrogate.debugDoingThisLayout);
+  }
+  expect(!layoutSurrogate.debugNeedsLayout || layoutSurrogate.debugDoingThisLayout, true);
+}
+
+void main() {
+  testWidgets('The overlay child sees the right inherited widgets', (WidgetTester tester) async {
+    int buildCount = 0;
+    TextDirection? directionSeenByOverlayChild;
+    TextDirection textDirection = TextDirection.rtl;
+    late StateSetter setState;
+
+    await tester.pumpWidget(
+      Directionality(
+        textDirection: TextDirection.ltr,
+        child: Overlay(
+          initialEntries: <OverlayEntry>[
+            OverlayEntry(
+              builder: (BuildContext context) {
+                return StatefulBuilder(
+                  builder: (BuildContext context, StateSetter setter) {
+                    setState = setter;
+                    return Directionality(
+                      textDirection: textDirection,
+                      child: OverlayPortal(
+                        overlayInfo: OverlayInfo.of(context),
+                        overlayChild: Builder(builder: (BuildContext context) {
+                          buildCount += 1;
+                          directionSeenByOverlayChild = Directionality.maybeOf(context);
+                          return const SizedBox();
+                        }),
+                        child: const SizedBox(),
+                      ),
+                    );
+                  }
+                );
+              },
+            ),
+          ],
+        ),
+      ),
+    );
+    expect(buildCount, 1);
+    expect(directionSeenByOverlayChild, textDirection);
+
+    setState(() {
+      textDirection = TextDirection.ltr;
+    });
+    await tester.pump();
+    expect(buildCount, 2);
+    expect(directionSeenByOverlayChild, textDirection);
+  });
+
+  testWidgets('Can reparent Overlay', (WidgetTester tester) async {
```

Comment:
in this test, parent of the overlay never really changes: The parent of the overlay is always the same keyed directionality widget.

Reply:
renamed.